crates/vm/src/arch/execution_control.rs

```diff
@@ -0,0 +1,169 @@
+use openvm_instructions::instruction::Instruction;
+use openvm_stark_backend::p3_field::PrimeField32;
+
+use super::{segment::VmExecutionState, ExecutionError, TracegenCtx, VmChipComplex, VmConfig};
+use crate::{
+    arch::{ExecutionState, InstructionExecutor},
+    system::memory::{online::GuestMemory, AddressMap, MemoryImage, PAGE_SIZE},
+};
+
+/// Check segment every 100 instructions.
+const SEGMENT_CHECK_INTERVAL: usize = 100;
+
+/// Trait for execution control, determining segmentation and stopping conditions
+pub trait ExecutionControl<F, VC>
+where
+    F: PrimeField32,
+    VC: VmConfig<F>,
+{
+    /// Guest memory type
+    type Mem: GuestMemory;
```

Comment:
what's the reasoning to have these as associated types?

Reply:
we can make them generic too. the reasoning to make these associated types is because i figured that the struct implementing the trait will usually implement it with a specific `Mem` and `Ctx` so the trait would never have to be implemented more than once for a struct and doesn't need to generic